pkg/resourcemanager/webhook/systemcomponentsconfig/handler.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package systemcomponentsconfig
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/gardener/gardener/pkg/utils"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	gsets "github.com/gardener/gardener/pkg/utils/sets"
+)
+
+// Handler contains required nodeSelector and tolerations information.
+type Handler struct {
+	Logger       logr.Logger
+	TargetClient client.Client
+
+	// NodeSelector is the selector used to retrieve nodes considered when calculating the tolerations.
+	NodeSelector map[string]string
+	// PodNodeSelector are the key-value pairs that should be added to each pod.
+	PodNodeSelector map[string]string
+	// PodTolerations are the tolerations that should be added to each pod.
+	PodTolerations []corev1.Toleration
+}
+
+// Default sets the spec.nodeSelector and spec.Tolerations fields the podTemplate.
+func (h *Handler) Default(ctx context.Context, obj runtime.Object) error {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return fmt.Errorf("expected *corev1.Pod but got %T", obj)
+	}
+
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	log := h.Logger.WithValues("pod", kutil.ObjectKeyForCreateWebhooks(pod, req))
+
+	if kutil.PodManagedByDaemonSet(pod) {
+		log.Info("Pod is managed by DaemonSet, skipping further handling")
+		return nil
+	}
+
+	log.Info("Handle node selector and system component tolerations")
+
+	// Add configured node selectors to pod.
+	h.handleNodeSelector(pod)
+
+	// Add tolerations for workers which allow system components.
+	return h.handleTolerations(ctx, log, pod)
+}
+
+func (h *Handler) handleNodeSelector(pod *corev1.Pod) {
+	pod.Spec.NodeSelector = utils.MergeStringMaps(pod.Spec.NodeSelector, h.PodNodeSelector)
+}
+
+func (h *Handler) handleTolerations(ctx context.Context, log logr.Logger, pod *corev1.Pod) error {
+	nodeList := &corev1.NodeList{}
+	if err := h.TargetClient.List(ctx, nodeList, client.MatchingLabels(h.NodeSelector)); err != nil {
+		return err
+	}
+
+	tolerations := gsets.New[corev1.Toleration]()
```

Comment:
`Toleration` has exactly one pointer field which is `tolerationSeconds` and today it's not used at all in the handler because we don't set this field when `Toleration`s are calculated from `Taint`s.
However, I see that this potentially can entail problems in the future. I will revise this accordingly.

Reply:
Please see 1fa46def92098d3bd3eb5737496ef99c067efc57. I like the new generic from API-Machinery but it's only based on the `comparable` constraint and thus have no room for any modifications. In `pkg/utils/sets` is a modified version of this set. Please let me know what you think and if you like it. I will add unit tests as soon as we agreed on the approach.

Reply:
Generally, this approach should be fine. TBH, I still find generics and their constraints a bit confusing and hard to read, though.

There are probably more readable ways to achieve the same, however with more lines of code.
E.g, using a fully-comparable struct type as the set element (which requires back-and-forth conversion), or using a map of hash strings to tolerations (no generics, only requires a hash function and a manual `UnsortedList` equivalent).
However, I'm fine with either approach you choose. You can decide :)

Reply:
Thanks for the feedback. I didn't like too much either, also due to the fact that we need to maintain our own set which is only required for special cases. I rather decided for a "fully-comparable struct" variant, see 7b8e8dbfd771eb1021a0e996bb8c578551587e8a.

Reply:
Ah nice, I haven't even thought about your approach. But it's even simpler than what I had in mind, I like it!